From 8c3286e71ac83c66c4db1497f6006a05ed5eb950 Mon Sep 17 00:00:00 2001
From: Herbert Xu <herbert@gondor.apana.org.au>
Date: Thu, 27 May 2010 11:50:19 +0800
Subject: [VAR] Do not poplocalvars prematurely on regular utilities

The recent cmdenviron removal broke regular utilities by calling
poplocalvars too early.  This patch fixes that by postponing the
poplocalvars for regular utilities until they have completed.

In order to ensure that local still works, it is now a special
built-in.

Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
(cherry picked from commit 1d806ac1fbafb867f6252e184e1be05c0829ab71)

Signed-off-by: Jonathan Nieder <jrnieder@gmail.com>
---
 ChangeLog           |    1 +
 src/builtins.def.in |    2 +-
 src/eval.c          |    8 +++++---
 3 files changed, 7 insertions(+), 4 deletions(-)

diff --git a/ChangeLog b/ChangeLog
index d29f0e4..d42b7d7 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,6 +1,7 @@
 2010-05-27  Herbert Xu <herbert@gondor.apana.org.au>
 
 	* Fix poplocalvar on abnormal exit from function.
+	* Do not poplocalvars prematurely on regular utilities.
 
 2010-05-26  Herbert Xu <herbert@gondor.apana.org.au>
 
diff --git a/src/builtins.def.in b/src/builtins.def.in
index 266d0ec..4441fe4 100644
--- a/src/builtins.def.in
+++ b/src/builtins.def.in
@@ -71,7 +71,7 @@ falsecmd	-u false
 getoptscmd	-u getopts
 hashcmd		hash
 jobscmd		-u jobs
-localcmd	-a local
+localcmd	-as local
 printfcmd	printf
 pwdcmd		pwd
 readcmd		-u read
diff --git a/src/eval.c b/src/eval.c
index 2cd931b..337667f 100644
--- a/src/eval.c
+++ b/src/eval.c
@@ -847,9 +847,11 @@ bail:
 		/* NOTREACHED */
 
 	case CMDBUILTIN:
-		poplocalvars(spclbltin > 0 || argc == 0);
-		if (execcmd && argc > 1)
-			listsetvar(varlist.list, VEXPORT);
+		if (spclbltin > 0 || argc == 0) {
+			poplocalvars(1);
+			if (execcmd && argc > 1)
+				listsetvar(varlist.list, VEXPORT);
+		}
 		if (evalbltin(cmdentry.u.cmd, argc, argv, flags)) {
 			int status;
 			int i;
-- 
1.7.4.1

